### PR TITLE
fix: components schema key names

### LINKED ISSUE #2804: bug: 2.4 changes schema key names
### Summary

There was a change in this PR (https://github.com/litestar-org/litestar/commit/219355c0e322fab41c921c56ca97d356b7ef868b#diff-6cdcc2ab6bdeda3d72588f3b65db0c437f7ba9cad43892829e75c4d701f9e658R659-R661) that updated the way schema keys are generated for the OpenAPI schema.

Where we originally had something like `UserAccount` as the model name, we know have `app_domain_accounts_UserAccount` or something similar.

### Basic Example

MCVE

```python
import json

from litestar import Litestar, get
from litestar.testing import TestClient
from pydantic import BaseModel


class SampleOne(BaseModel):
    property_one: str


class SampleTwo(BaseModel):
    __module__ = ""
    property_one: str


@get("/one")
async def test_handler_one() -> SampleOne:
    return SampleOne(property_one="test")


@get("/two")
async def test_handler_two() -> SampleTwo:
    return SampleTwo(property_one="test")


app = Litestar(route_handlers=[test_handler_one, test_handler_two], debug=True)

with TestClient(app=app) as client:
    response = client.get("/schema/openapi.json")
    print(json.dumps(response.json()["components"]["schemas"], indent=2))
```

This outputs:

```
{
  "main_SampleOne": {
    "properties": {
      "property_one": {
        "type": "string"
      }
    },
    "type": "object",
    "required": [
      "property_one"
    ],
    "title": "SampleOne"
  },
  "SampleTwo": {
    "properties": {
      "property_one": {
        "type": "string"
      }
    },
    "type": "object",
    "required": [
      "property_one"
    ],
    "title": "SampleTwo"
  }
}INFO - 2023-11-29 18:43:59,780 - httpx - _client - HTTP Request: GET http://testserver.local/schema/openapi.json "HTTP/1.1 200 OK"
```

Notice that for the `SampleOne` model the generated schema key name is `main_SampleOne`.

After reviewing the PR as far as I can understand it you are now making use of class metadata to get name of the model, whereas previously you were using the schema title field.

The only workaround I've been able to come up with can be seen in the the `SampleTwo` model where I set `__module__ = ""`. This allows me to get only the class name as the schema key as can be seen in the output.

### Drawbacks and Impact

This is a breaking change for us as we use an OpenAPI typescript generator from the generated `openapi.json`
and the named types are now different than what it expects.

### Unresolved questions

1. Is there any other way to define the model name other than the workaround presented here
2. If you take on this feature request then the implementation details are up to you but perhaps this could be achieved with a new dunder method on the class and if it's there, use it otherwise use the `__module__.__name__`

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh Litestar dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/2804">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/2804/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by peterschutt**
### Pull Request Checklist

- [x] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR
- [x] Pre-Commit Checks were ran and passed
- [x] Tests were ran and passed

### Description

https://github.com/litestar-org/litestar/commit/219355c0e322fab41c921c56ca97d356b7ef868b#diff-6cdcc2ab6bdeda3d72588f3b65db0c437f7ba9cad43892829e75c4d701f9e658R659-R661 introduced a change to the way that keys for the openapi components/schemas object were calculated to address name the possibility of name collisions.

This PR reverts that behavior for the case where a name has no collision, and only introduces extended keys for the case where there are multiple objects with the same name.

To do this, it replaces the pattern of carrying the `components/schemas` section of the openapi document around in the `OpenAPIContext` object, and replaces it with a `SchemaRegistry` object.

The schema registry tracks all schemas created for named objects, and any references to them. Once all schemas are known, we build the `components/schemas` object from the openapi doc, using the shortest possible paths to avoid collision.

### Close Issue(s)

Closes #2804

### CODE REVIEW

File: litestar/_openapi/datastructures.py [hunk 1]
```diff
@@ -1,39 +1,99 @@
...
+
+        return components_schemas
+
+
+class OpenAPIContext:
+    def __init__(
+        self,
+        openapi_config: OpenAPIConfig,
+        plugins: Sequence[OpenAPISchemaPluginProtocol],
+        schemas: dict[str, Schema],
```
Comment (peterschutt) [inline litestar/_openapi/datastructures.py:90]: would like to try and replace `schemas` with `schema_registry` altogether, but not sure if possible yet.

File: tests/unit/test_contrib/test_repository.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 
 
 def test_app_repository_signature_namespace() -> None:
-    app = Litestar([], on_app_init=[handlers.on_app_init])
+    app = Litestar([], on_app_init=[handlers.on_app_init], openapi_config=None)
```
Comment (peterschutt) [inline tests/unit/test_contrib/test_repository.py:19]:
```suggestion
    app = Litestar([], on_app_init=[handlers.on_app_init])
```

**Comment by JacobCoffee** [comment]
Ill be testing this in the morning if we can hold of on merging in the middle of the night :P

**Comment by peterschutt** [comment]
> Ill be testing this in the morning if we can hold of on merging in the middle of the night :P

Thanks! No rush to merge from me.

**Review by JacobCoffee — APPROVED** [review]
worked well in testing on private apps

**Pull request merged**